### PR TITLE
Add support for intersection type

### PR DESCRIPTION
**Pull request by Maninda**
## Purpose
Related to https://github.com/ballerina-platform/ballerina-standard-library/issues/2619

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [x] Added tests

### CODE REVIEW

File: compiler-plugin/src/main/java/io/ballerina/stdlib/email/compiler/EmailServiceValidator.java [hunk 1]
```diff
@@ -248,13 +251,21 @@ private boolean hasNoParameters(SeparatedNodeList<ParameterNode> parameterNodes,
...
         for (ParameterNode parameterNode : parameterNodes) {
             RequiredParameterNode requiredParameterNode = (RequiredParameterNode) parameterNode;
             Node parameterTypeName = requiredParameterNode.typeName();
             DiagnosticInfo diagnosticInfo;
-            if (functionName.equals(ON_MESSAGE) && (!parameterTypeName.toString().contains(EMAIL_MESSAGE)
-                    || !parameterTypeName.toString().contains(MODULE_NAME))) {
+            if (functionName.equals(ON_MESSAGE) && !(inputParams.get(0).typeDescriptor().signature().
+                    equals(moduleId + ":" + EMAIL_MESSAGE) || (inputParams.get(0).typeDescriptor().signature().
+                    contains(EMAIL_MESSAGE) && inputParams.get(0).typeDescriptor().typeKind() ==
+                    TypeDescKind.INTERSECTION))) {
```
Comment (ayeshLK) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/email/compiler/EmailServiceValidator.java:268]: Shall we extract this validation into a another method ?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/email/compiler/EmailServiceValidator.java [hunk 2]
```diff
@@ -248,13 +251,21 @@ private boolean hasNoParameters(SeparatedNodeList<ParameterNode> parameterNodes,
...
+        List<ParameterSymbol> inputParams = functionTypeSymbol.params().get();
+        String moduleId = getModuleId(inputParams.get(0));
+
         for (ParameterNode parameterNode : parameterNodes) {
             RequiredParameterNode requiredParameterNode = (RequiredParameterNode) parameterNode;
             Node parameterTypeName = requiredParameterNode.typeName();
             DiagnosticInfo diagnosticInfo;
-            if (functionName.equals(ON_MESSAGE) && (!parameterTypeName.toString().contains(EMAIL_MESSAGE)
-                    || !parameterTypeName.toString().contains(MODULE_NAME))) {
+            if (functionName.equals(ON_MESSAGE) && !(inputParams.get(0).typeDescriptor().signature().
```
Comment (ayeshLK) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/email/compiler/EmailServiceValidator.java:265]: Lets say the developer has declared the service as follows:
```ballerina
import ballerina/email as em;

service "testReadonlyPopService" on new em:PopListener({
                                    host: "pop.example.com",
                                    username: "abc@example.com",
                                    password: "pass123",
                                    pollingInterval: 2,
                                    port: 995
                                }) {

    remote function onMessage(readonly & em:Message emailMessage) {

    }
}
```
In that case will `inputParams.get(0).typeDescriptor().signature()` give the correct type description ?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/email/compiler/EmailServiceValidator.java (hunk 1 above)
Comment (Maninda) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/email/compiler/EmailServiceValidator.java]: Extracted.

**Review by ayeshLK — APPROVED** [review]
LGTM

**Comment by Bhashinee** [comment]
Just wondering, don't we need to do any changes in the native code where the dispatching happens? For example, creating the `Message emailMessage` as read-only in the native code and then dispatch to the `onMessage`?

**Comment by shafreenAnfar** [comment]
@Bhashinee yeah, exactly. I was wondering the same. It seems this PR only fix the complier plugin.

**Comment by Maninda** [comment]
> @Bhashinee yeah, exactly. I was wondering the same. It seems this PR only fix the complier plugin.

Changed.

**Comment by Maninda** [comment]
As discussed offline with Shafreen we are closing this PR and instead remove the acceptance of `readonly` type from the listener as given by https://github.com/ballerina-platform/ballerina-standard-library/issues/2633 .

**Pull request closed without merging**